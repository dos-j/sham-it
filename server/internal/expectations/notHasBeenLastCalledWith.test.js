const notHasBeenLastCalledWith = require("./notHasBeenLastCalledWith");

describe("unit: notHasBeenLastCalledWith", () => {
  let handler;
  let requestStore;

  beforeEach(() => {
    requestStore = [];
    handler = notHasBeenLastCalledWith(requestStore);
  });

  test("it should return 204 No Content if the query given matches the number of requests in the requestStore", () => {
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test1",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test2",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test3",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });

    expect(handler({ op: "==", prop: "pathname", value: "/test2" })).toEqual({
      status: 204
    });
  });

  test("it should return 204 No Content if there is no last request", () => {
    expect(handler({ op: "==", prop: "pathname", value: "/test2" })).toEqual(
      expect.objectContaining({
        status: 204
      })
    );
  });

  test("it should return 417 Expectation failed if query given does not match the number of requests in the requestStore", () => {
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test1",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test2",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });

    expect(handler({ op: "==", prop: "pathname", value: "/test2" })).toEqual(
      expect.objectContaining({
        status: 417
      })
    );
  });

  test("it should show a detailed error message including the query and the last request", () => {
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test1",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test2",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });
    requestStore.push({
      request: {
        method: "GET",
        pathname: "/test3",
        query: { a: 1234 },
        headers: { accept: "application/json" }
      }
    });

    expect(
      handler({ op: "==", prop: "pathname", value: "/test3" })
    ).toMatchSnapshot();
  });
});
