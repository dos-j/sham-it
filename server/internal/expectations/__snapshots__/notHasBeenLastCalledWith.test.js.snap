// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit: notHasBeenLastCalledWith it should show a detailed error message including the query and the last request 1`] = `
Object {
  "body": "Expected sham not to have been last called with:
  {
    \\"op\\": \\"==\\",
    \\"prop\\": \\"pathname\\",
    \\"value\\": \\"/test3\\"
  }
But it was last called with:
  GET /test3?a=1234
    accept: application/json",
  "status": 417,
}
`;
